framework/sandstone.cpp

```diff
@@ -1018,6 +1019,10 @@ static uintptr_t thread_runner(int thread_number)
 
     cleanup.run_now();
 
+    // If rescheduling, do cleanup
+    if (sApp->device_schedule != nullptr)
+        sApp->device_schedule->finish_reschedule();
```

Comment:
This needs to be inside of the cleanup lambda, so it will be executed even if the test function throws. Put it after the `test_end` call, so it's after the `logging_mark_thread_failed()` and `ud2()` calls too.